Graph generation (#401)
ros-navigation/docs.nav2.org: tutorials/docs/route_graph_generation.rst

transformation.

|

 .. image:: images/route_graph_generation/georeferencer.png
    :height: 807px
    :width: 1460px
    :align: center

|

Drag and drop the raster file output into the layers window.

|

 .. image:: images/route_graph_generation/raster_layer.png
    :height: 702px
    :width: 1051px
    :align: center

|

4- Now that we have the raster layer in the correct coordinate system we can start placing nodes.

Select `Layer -> Create Layer -> New ShapeFile Layer`. Set the shapefile layer setting to be 
`nodes` for the `File name`, `points` for the `Geometry type` and `WGS 84/ Pseudo-Mercator` for the coordinate system. Press `OK`. 

|

 .. image:: images/route_graph_generation/node_layer.png
    :height: 1041px
    :width: 887px
    :align: center

|

In order to have the `id` field auto increment, right click on the layer and select the `Attribute Form`. 
Expand the `Fields` drop down menu and select `id`. Select the `Expression Dialog` that is next to the `Default value`. 

|

 .. image:: images/route_graph_generation/attribute_form.png
    :height: 788px
    :width: 1107px
    :align: center

|

> Within the `Expression Dialog` select `Import user expressions` and import `increment_node_id.json`. Expand `User expressions` and double click on `increment_node_id`. 

Reviewer: Where is `increment_node_id.json`, where do I get it? Where does it live?

Reviewer: Same with any of these new files / info

.. _route_graph_generation: 

Route Graph Generation
**********************

- `Overview`_
- `Requirements`_
- `Tutorial Steps`_

Overview
========
This tutorial walks a user through generating a graph for the nav2 route server.

Requirements
============
Follow https://www.qgis.org/en/site/forusers/download.html to install QGIS.

Tutorial Steps
==============

> 1- Open QGIS and create a new project by selecting `Project->New`. Save your project by selecting, `Project -> Save As`. 

Reviewer: I didn't mention this before since its a final cleanup thing but we're at this point now: For all the numbered steps, make them subsections. If you use the `-----` header type, then it'll fall under the Tutorial Steps section but be more easily pronounced.

So like 
```
1- Create New QGIS Project
----------------------------------
```

or 

```
2- Setup Coordinate System
-------------------------------------
```

To georeference the image, three control points will choosen. Each control point maps a pixel location to the spatial reference. 

For example, let's say you have a 100 by 200 raster image with a resolution of 0.05 and a origin in the bottom left corner.
The first control point we could select is (0, -200)px -> (0, 0)m. (0, -200)px is the bottom left corner in the pixel frame and (0, 0)m is coordinates in the spatial frame. 
For the second control point we will select the top left corner of the image as the original (0, 0)px. The associated coordinates in the spatial frame can be calculated by using the resolution 
to convert the pixels to meters. In this case, (0, 0)px -> (0, 10)m. Where 10 m is 200 px* 0.05 m/px. The last control will be the bottom right corner: (200, -200)px -> (5, 0)m. 

To georeference image select `Raster -> Georefencer`. Set the `Trasformation Settings` to `Linear` for `Transformation 
type`, `WGS 84/ Pseudo-Mercator` for `Target SRS` and set your desired path for the `Output Raster`. 

|

 .. image:: images/route_graph_generation/transformation_settings.png
    :height: 757px
    :width: 458px
    :align: center

|

> Select the raster image you wish to georeference and place control point by selecting `Add Point ` and clicking on the raster image. 

Reviewer: ```suggestion
Select the raster image you wish to georeference and place control point by selecting `Add Point` and clicking on the raster image. 
```

I think this caused some rendering problems

Set the project coordinate reference system by selecting `Project->Properties->CRS`. Set the coordinate system to `WGS 84/ Pseudo-Mercator` and click `OK`. 

.. note:: 
    The `WGS 84/ Pseudo-Mercator` is one of few coordinate systems that could be used. 
    For more information about coordinate reference systems please review https://docs.qgis.org/3.22/en/docs/gentle_gis_introduction/coordinate_reference_systems.html. 

|

 .. image:: images/route_graph_generation/coordinate_reference_system.png
    :height: 1097px
    :width: 1064px
    :align: center

|

2- By default, the imported raster image origin will be at the top left corner, with x pointing to the right, and y pointing up.
The size of the image in meters is equal to the equal to the size in pixels.
To correctly set the orgin and the resolution we will georeference the raster image. 
To georeference the image, three control points will choosen. Each control point maps a pixel location to the spatial reference. 

For example, let's say you have a 100 by 200 raster image with a resolution of 0.05 and a origin in the bottom left corner.

> The first control point we could select is (0, -200) -> (0,0). (0, -200) is the bottom left corner in the pixel frame and (0, 0) is coordinates in the spatial frame. 

Reviewer: Is that an instruction to set `(0, -200)` to `(0,0)`? Make it clear that you're asking them to do something. Is that automatically set / detected? Or do they have to type in (0,0)?

Reviewer: Make it clear we're setting it into meters space, not pixel coordinate space. 0,0 is ambiguous for units and the later ones are talking about meters so it would be good to clarify that up front

> For the second control point we will select the top left corner of the image (0,0). The associated coordinates in the spatial frame can be calculated by using the resolution 

Reviewer: ```suggestion
For the second control point we will select the top left corner of the image as the original (0,0). The associated coordinates in the spatial frame can be calculated by using the resolution 
```

> to convert the pixels to meters. In this case, (0, 0) -> (0, 10). Where 10 m is 200 * 0.05. The last control will be the bottom right corner. (200,-200) -> (5, 0). 

Reviewer: ```suggestion
to convert the pixels to meters. In this case, (0, 0) -> (0, 10). Where 10 m is 200 * 0.05. The last control will be the bottom right corner: (200,-200) -> (5, 0). 
```

> .. note:: 

Reviewer: Move this before the example

    Georeferencing is type of coordinate transformation that relates a digital raster image to a spatial reference. 

To georeference image select `Raster -> Georefencer`. Set the `Trasformation Settings` to `Linear` for `Transformation 
type`, `WGS 84/ Pseudo-Mercator` for `Traget SRS` and set your desired path for the `Output Raster`. 

|

 .. image:: images/route_graph_generation/transformation_settings.png
    :height: 757px
    :width: 458px
    :align: center

|

> Select the raster image you wish to georeference and place control point by selecting the icon next to the `Transformation Settings` and clicking on the raster image. 

Reviewer: Icon of what? Where - what screen?

The control points can be modified by clicking on the control point table and typing in new values. Once you are satified with your control points, apply the transformation

> by selecting `Start Transformation`. 

Reviewer: What button? does it lierally say "start transformation"? I don't see that on the images provided. If there's an icon, show me or tell me what it is (or have an arrow pointing to it in the graphic)

|

 .. image:: images/route_graph_generation/node_layer.png
    :height: 1041px
    :width: 887px
    :align: center

|

In order to have the `id` field auto increment, right click on the layer and select the `Properties -> Attribute Form`. 
Expand the `Fields` drop down menu and select `id`. Select the `Expression Dialog` icon which is next to the `Default value` field. 

|

 .. image:: images/route_graph_generation/attribute_form.png
    :height: 788px
    :width: 1107px
    :align: center

|

Within the `Expression Dialog` select `Import user expressions` and import `increment_node_id.json` which exists in the scripts file with the nav2 route module. Expand `User expressions` and double click on `increment_node_id`. 
The expression should show up in the left window. Click `OK` to save the expression and exit the `Expression Dialog`. Then click `Apply` and `OK` and save and exit the `Attributes Form`.
This will increment the node `id` by one every time a new node is added. The first node `id` will be zero. 

|

 .. image:: images/route_graph_generation/expression_dialog.png
    :height: 776px
    :width: 950px
    :align: center

|

> Click on the node layer and then on the pencil icon to start editing the layer. To add points, click on `Add points feature`. Start adding nodes by clicking in the main window.

Author: I described the icon. If that doesn't work I will add an arrow.

> Click on the node layer and then on the pencil icon to start editing the layer. To add points, click on `Add points feature`. Start adding nodes by clicking in the main window.

Reviewer: Ok... but you should still actually tell the users what the button is called... I think you should keep the text in line as you had before and add an arrow to the image to show where it is, that's the most clear.

    or multiple nodes are deleted. Any node that is added will contiune to increment based on the highest id. This value can be overwriten by the user 
    though the pop window after node placement or though the `Attribute Table`. To delete features in a layer right click on the layer and select `Open Attribute Table`. 
    Click on `Target Multi Edit Mode`, select the features to delete, and click on the trash can icon.     

4- Now an edge layer can be created. Select `Layer -> Create Layer -> New ShapeFile Layer`. Set the shapefile layer setting to be 
`edges` for the `File name`, `LineString` for the `Geometry type` and `WGS 84/ Pseudo-Mercator` for the coordinate system. Press `OK`.
To auto incrment the `id` field follow the same steps as above except replace `increment_node_id.json` with `increment_edge_id.json`. 
This will increment the edge `id` by one every time a new edge is added. The first edge `id` will be `10000`. 

|

 .. image:: images/route_graph_generation/edge_layer.png
    :height: 1041px
    :width: 887px
    :align: center

|

> Click on the edge layer and then on the pencil icon to start editing the layer. To add edges, click on `Add Line feature`. Start adding edges by clicking twice in the main window. 

Author: same as above.

> Click on the edge layer and then on the pencil icon to start editing the layer. To add edges, click on `Add Line feature`. Start adding edges by clicking twice in the main window. 

Reviewer: Ok... but you should still actually tell the users what the button is called...  I think you should keep the text in line as you had before and add an arrow to the image to show where it is, that's the most clear.

(First point is start, second point is end). Press `Esc` when you have finished adding the two points.  

|

 .. image:: images/route_graph_generation/edges.png

>     :height: 1922px

Author: yeahhhh fixed

    :width: 1082px
    :align: center

|

.. note:: 

>     To see the directionality of the edges, right click on the edge layer and select `Properties`. Click `Symbology -> Simple Line` and change the `Symbol layer type` to `Arrow`.

Reviewer: Actually, don't have this be a footnote later, have them set this up, up front . Its important to know directionality since there are no bi-directional edges.

Author: I moved this up.

For the second control point we will select the top left corner of the image as the original (0, 0)px. The associated coordinates in the spatial frame can be calculated by using the resolution 
to convert the pixels to meters. In this case, (0, 0)px -> (0, 10)m. Where 10 m is 200 px* 0.05 m/px. The last control will be the bottom right corner: (200, -200)px -> (5, 0)m. 

To georeference image select `Raster -> Georefencer`. Set the `Trasformation Settings` to `Linear` for `Transformation 
type`, `WGS 84/ Pseudo-Mercator` for `Target SRS` and set your desired path for the `Output Raster`. 

|

 .. image:: images/route_graph_generation/transformation_settings.png
    :height: 757px
    :width: 458px
    :align: center

|

Select the raster image you wish to georeference and place control point by selecting `Add Point` and clicking on the raster image. 
The control points can be modified by clicking on the control point table and typing in new values. Once you are satified with your control points, apply the transformation
by selecting `Start Transformation`. 

|

 .. image:: images/route_graph_generation/georeferencer.png
    :height: 400px
    :width: 730px
    :align: center

|

> Drag and drop the georeferenced raster file into the layers window.

Reviewer: ```suggestion
Close the window. Drag and drop the georeferenced raster file that was just created into the layers window. This should be a tiff file, by default.
```

`edges` for the `File name`, `LineString` for the `Geometry type` and `WGS 84/ Pseudo-Mercator` for the coordinate system. Press `OK`.
To auto incrment the `id` field follow the same steps as above except replace `increment_node_id.json` with `increment_edge_id.json`. 
This will increment the edge `id` by one every time a new edge is added. The first edge `id` will be `10000`. 

|

 .. image:: images/route_graph_generation/edge_layer.png
    :height: 520px
    :width: 440px
    :align: center

|

To see the directionality of the edges, right click on the edge layer and select `Properties`. Click `Symbology -> Simple Line` and change the `Symbol layer type` to `Arrow`.
Click `Ok`. You should now be able to see arrows for each line string. Note that all edges in the route graph are directional, so if you want a bidirectional connection you should duplicate it in reverse.

Click on the edge layer and then on the `Toggle Editing` icon represented as a pencil to start editing the layer. 
To add edges, click on the `Add Line Feature`. Start adding edges by clicking twice in the main window. 

> (First point is start, second point is end). Press `Esc` when you have finished adding the two points.  

Reviewer: ```suggestion
(First point is start, second point is end). Press right click anywhere when you have finished adding the two points for the prompt to save to come up.
The edge locations must exactly match up to the node locations, so if Snapping is not already enabled by default, enable it via Views->Toolbars->Snapping Toolbar and depress the Magnet icon.
```
